### PR TITLE
Add functools.reduce

### PR DESCRIPTION
**Pull request by mpajkowski**

### CODE REVIEW

**Comment by palaviv** [comment]
@mpajkowski why have you chosen to implement this in Rust instead of using functools.py from CPython stdlib?

**Comment by windelbouwman** [comment]
@palaviv CPython also implements `reduce` in C code, I think this is why.

**Comment by windelbouwman** [comment]
@mpajkowski nice job!

File: vm/src/stdlib/functools.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    vm: &VirtualMachine,
+) -> PyResult {
+    let iterator = objiter::get_iter(vm, &sequence)?;
+
+    let start_value = if let OptionalArg::Present(val) = start_value {
+        val
+    } else {
+        objiter::call_next(vm, &iterator).map_err(|err| {
+            if objtype::isinstance(&err, &vm.ctx.exceptions.stop_iteration) {
+                let exc_type = vm.ctx.exceptions.type_error.clone();
```
Comment (youknowone) [inline vm/src/stdlib/functools.rs]: `vm.new_type_error("reduce() of empty sequence with no initial value".to_string())` also works

**Pull request merged**